cluster-up/cluster/kind-k8s-sriov-1.17.0/sriov-node/node.sh

```diff
@@ -0,0 +1,108 @@
+#!/bin/bash
+
+SCRIPT_PATH=${SCRIPT_PATH:-$(dirname "$(realpath "$0")")}
+
+CONFIGURE_VFS_SCRIPT_PATH="${SCRIPT_PATH}/configure_vfs.sh"
+PFS_IN_USE=${PFS_IN_USE:-}
+
+function node::discover_host_pfs() {
+  local -r sriov_pfs=( $(find /sys/class/net/*/device/sriov_numvfs) )
+  [ "${#sriov_pfs[@]}" -eq 0 ] && echo "FATAL: Could not find available sriov PFs on host" >&2 && return 1
+
+  local pf_name
+  local pf_names=()
+  for pf in "${sriov_pfs[@]}"; do
+    pf_name="${pf%%/device/*}"
+    pf_name="${pf_name##*/}"
+    if [ $(echo "${PF_BLACKLIST[@]}" | grep "${pf_name}") ]; then
+      continue
+    fi
+
+    pfs_names+=( $pf_name )
+  done
+
+  echo "${pfs_names[@]}"
+}
+
+# node::configure_sriov_pfs_and_vfs moves SRIOV PF's to nodes netns,
+# create SRIOV VF's and configure their driver on each node.
+# Exports 'PFS_IN_USE' env variable with the list of SRIOV PF's
+# that been moved to nodes netns.
+function node::configure_sriov_pfs_and_vfs() {
+  local -r nodes_array=($1)
+  local -r pfs_names_array=($2)
+  local -r pf_count_per_node=$3
+
+  local -r config_vf_script=$(basename "$CONFIGURE_VFS_SCRIPT_PATH")
+  local pfs_to_move=()
+  local pfs_array_offset=0
+  local pfs_in_use=()
+  local node_exec
+
+  # 'iplink' learns which network namespaces there are by checking /var/run/netns
+  mkdir -p /var/run/netns
+  for node in "${nodes_array[@]}"; do
+    prepare_node_netns "$node"
+
+    ## Move PF's to node netns
+    # Slice '$pfs_names_array' to have unique silce for each node
+    # with '$pf_count_per_node' PF's names
+    pfs_to_move=( "${pfs_names_array[@]:$pfs_array_offset:$pf_count_per_node}" )
+    echo "Moving '${pfs_to_move[*]}' PF's to '$node' netns"
+    for pf_name in "${pfs_to_move[@]}"; do
+      move_pf_to_node_netns "$node" "$pf_name"
+    done
+    # Increment the offset for next slice
+    pfs_array_offset=$((pfs_array_offset + pf_count_per_node))
+    pfs_in_use+=( $pf_name )
+
+    # KIND mounts sysfs as read-only by default, remount as R/W"
+    node_exec="docker exec $node"
+    $node_exec mount -o remount,rw /sys
+    $node_exec chmod 666 /dev/vfio/vfio
+
+    # Create and configure SRIOV Virtual Functions on SRIOV node
+    docker cp "$CONFIGURE_VFS_SCRIPT_PATH" "$node:/"
+    $node_exec bash -c "DRIVER=$VFS_DRIVER DRIVER_KMODULE=$VFS_DRIVER_KMODULE ./$config_vf_script"
+
+    _kubectl label node $node $SRIOV_NODE_LABEL
+  done
+
+  export PFS_IN_USE="${pfs_in_use[@]}"
```

Comment:
Why suddenly its needed?

Reply:
We need this for the caller to pass the correct PF names to `sriov_components::deploy` which patch the sriodp config (`SriovNetworkNodePolicy` equivalent)
https://github.com/kubevirt/kubevirtci/blob/2a7cdf82fcb91af08892b58720a01a2597458485/cluster-up/cluster/kind-k8s-sriov-1.17.0/config_sriov_cluster.sh#L52-L62

Reply:
Ok thanks

One point, why not to just allocate the PFs to the node,
and later peek into the node and then see what PFs it has ?
as we had it before.
It will simplifies the flow and maintenance of that,
is there any reason to not do that?

I am fine to do it on next PRs if there is an agreement that it should be done.

Reply:
I am not sure I understand, with this PR changes it checks what PFs are available on the host
then it will move PF's to each worker node according to `PFS_COUNT_PER_NODE`.

We need to support a setup with multiple nodes where each node could have more then one PF
and also consider multi cluster creation.
This PR add more layers to support this not trivial setup with all its constrains.

Reply:
Multi cluster has two variants, local and on CI, lets put them aside for a moment, since we can make them agnostic for the changes almost.

I suggest that we first move to each worker node PFs according PFS_COUNT_PER_NODE.
Then you don't need to export `PFS_IN_USE`
but just peek in the nodes, and see how many PFs it has and which, once you count the VF and config the VFIO.

The additional layer i meant that isn't needed imo, is the discover, which is decoupled now from moving the PF into the node and so on.

Reply:
> I suggest that we first move to each worker node PFs according PFS_COUNT_PER_NODE.
> Then you don't need to export `PFS_IN_USE`
> but just peek in the nodes, and see how many PFs it has and which, once you count the VF and config the VFIO.
> 
> The additional layer i meant that isn't needed imo, is the discover, which is decoupled now from moving the PF into the node and so on.

Doing all (discover PFs, move to nodes, configure VF's) on the same function was too much so I decided to split it.

When moving the PF's to nodes, they are saved in variable that is used later for patching sriov-dp config
I think its better then moving them and then check the nodes to get them, we already have the names when we move them.

I cant use the variable that ` node::configure_sriov_pfs_and_vfs ` returns because it will include all the stdout content it generates.

There are few possible solutions:
1. Remove the export keyword, its could work w/o it because we source node.sh script 
    so eventually its the same context.
2. Define `PFS_IN_USE` variable from `node::configure_sriov_pfs_and_vfs` on the fly by, passing the name of the variable
    (@EdDev recommendation)
3. If adding another variable is unbearable, I can split `node::configure_sriov_pfs_and_vfs` in a way that we could
    use returned variable with the used PF's
    Something like:
    ```
    PFS_IN_USE=$(node::move_pfs $worker_nodes $pfs "$PF_COUNT_PER_NODE")
    node::configure_vfs $worker_nodes
    node::label_sriov_capable $worker_nodes
    ```

> Again, its not needed to be in this PR, unless we decide its needed, but first lets please sync on the idea.

Cool

Reply:
> Doing all (discover PFs, move to nodes, configure VF's) on the same function was too much so I decided to split it.

I am looking on it otherwise:
1. You move the PF to the node, and check if it moved.
2. You create VF on the PFs that the node have.

The node has all the entropy it needs (all the information), no need additional middleman to convey the information.
Its the minimal operations that are needed anyhow.

Reply:
Yes I understand that all the info is there,
I split it to improve readability, it was already long in the first place and with this PR changes it got even longer
and hard to follow.

I picked option (2) on this PR, we can implement option (3) on a following PR

Reply:
You dont even need PFS_IN_USE in option 3, or in what i meant, the info is in the node itself,
unless we speak on the same thing